routemaster/services/throttle.rb

```diff
@@ -0,0 +1,106 @@
+require 'routemaster/services'
+require 'routemaster/mixins/log'
+
+# throttle |ˈθrɒt(ə)l|
+#   noun
+#   a device controlling the flow of fuel or power to
+#   an engine: the engines were at full throttle.
+#
+module Routemaster
+  module Services
+    class Throttle
+      MAX_HP = 100
+
+      def initialize(batch: nil, subscriber: nil)
+        @batch = batch
+        @subscriber = subscriber || @batch.subscriber
+      end
+
+      # Always continue if the backoff is per batch.
+      # Always continue if the Subscriber is perfectly healthy.
+      #
+      # If the Subscriber is not healthy, calculate what the
+      # current backoff delay would be. If the Subscriber hasn't
+      # been hit for an amount of time that exceeds the
+      # calculated backoff, then the Subscriber has already
+      # had enough time to recover and the delivery can be
+      # attempted imemediately.
+      #
+      def should_deliver?
+        return true if _strategy == :batch
+        return true if @subscriber.health_points >= MAX_HP
+
+        delay = _subscriber_backoff
+        last_attempt = @subscriber.last_attempted_at
+
+        _stale_enough?(last_attempt, delay)
+      end
+
+
+      def retry_backoff
+        if _strategy == :subscriber
+          _subscriber_backoff
+        else
+          _batch_backoff
+        end
+      end
+
+
+      private
+
+      # Is a timestamp older than a certain time interval?
+      #
+      def _stale_enough?(timestamp, time_span)
+        timestamp < (Routemaster.now - time_span)
+      end
+
+
+      def _subscriber_backoff
+        hp = @subscriber.health_points
+        _exponential_backoff(_health_to_severity(hp))
+      end
+
+      # Use incremental ranges (Fibonacci) to map a subscriber's
+      # health points to a number of hypotetical failed attempts.
```

Comment:
Basically you're proposing an exponential growth for the backoffs, while I implemented a logarithmic growth.
Happy to adopt your function if it will serve us better.